### PR TITLE
Create basePersistDirectory only when needed

### PR DESCRIPTION
**Pull request by praveev**
@gianm is this what you had in mind for https://github.com/druid-io/druid/issues/3347? 
I'm not sure how to repro and test the random tempDir creation failure tho.

### CODE REVIEW

**Comment by praveev** [comment]
hrm...thinking back. That fix may not be desirable. 
Is it alright if different temp dirs (when `basePersistDirectory` ==  null ) are created on multiple calls to `getBasePersistDirectory`?

**Comment by gianm** [comment]
@praveev sorry for the delay. Your hunch is right, we _don't_ want to create new temp dirs on each call to getBasePersistDirectory. The other tricky thing about this is that sometimes, things call getBasePersistDirectory even if they don't actually need to make a base persist directory.

One example where this happens is that KafkaTuningConfig gets defaults by doing `RealtimeTuningConfig.makeDefaultTuningConfig(basePersistDirectory)` and then calling get on the various methods, including getBasePersistDirectory. But it doesn't actually want to create the directory at that point. It might never create the directory, especially if withBasePersistDirectory is called to give it a new directory.

So what we want for #3347 is a scheme that allows the things that actually write to getBasePersistDirectory to get the behavior they need (if they call it multiple times, they should get the same response) but for things that _don't_ write to getBasePersistDirectory to avoid triggering needless directory creations.

**Comment by praveev** [comment]
@gianm does using builder pattern serve this purpose? Unit test seem to pass. This ensures that those who want basePersistDirectory explicitly create it ahead of time. No creation is done in the getBasePersistDirectory() call

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 1]
```diff
@@ -60,7 +60,9 @@ public KafkaTuningConfig(
   )
   {
     // Cannot be a static because default basePersistDirectory is unique per-instance
-    final RealtimeTuningConfig defaults = RealtimeTuningConfig.makeDefaultTuningConfig(basePersistDirectory);
+    final RealtimeTuningConfig defaults = new RealtimeTuningConfig.Builder()
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: This isn't quite right, but flows from a mistake in the original code. KafkaTuningConfig shouldn't have a `basePersistDirectory`. It's not documented and it doesn't make sense. (Base persist directory for tasks is set by runtime properties and always overridden via `withBasePersistDirectory`; so it shouldn't be provided in the task.)

Just remove basePersistDirectory from the constructor and let it stay null here.

**Comment by gianm** [comment]
@praveev are you able to pick this back up?

**Comment by praveev** [comment]
@gianm I am currently on vacation, I won't be able to get back to this until mid-April.

**Comment by gianm** [comment]
Got it, no worries, was just checking in. Looking forward to reviewing again when you get back 😄

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java (hunk 1 above)
Comment (praveev) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: hope i understood your comment right. i've updated the pr

**Comment by praveev** [comment]
@gianm Updated the PR. Not entirely sure if I understood your comment right, esp about the basePersistDirectory and KafkaTuningConfig. Please take a look and let me know.

**Comment by fjy** [comment]
@praveev can we finish this?

**Comment by praveev** [comment]
waiting on @gianm to review the updates

**Comment by gianm** [comment]
Sorry @praveev, I totally missed this. If you are still interested in working on this patch, could you please resolve the conflicts and then we could take another look.

**Comment by jihoonson** [comment]
The original issue (#3347) has been fixed in https://github.com/apache/incubator-druid/pull/5648, but this PR looks to solve another issue of postponing creating basePersistDirectory. Do we still need this? Anyway, I'll remove the milestone here.

**Pull request closed without merging**